### PR TITLE
Initial training and evaluation

### PR DESCRIPTION
**Pull request by jdiaz4302**
Adding two phases

* `02_train` which adds `A_FitBasicLSTM.ipynb`

The goal of this notebook is a basic, no-advanced-tricks LSTM that is properly/fully trained. The ultimate goal is this notebook/model be a comparison point for a decked out LSTM (to be completed later this week) and later transformer efforts. 

This notebook saves the final model predictions for the training and validation set as well as the data scaling factors and the model weights.

* `03_eval` which adds `A_EvaluationMetrics.ipynb` and `B_PlottingInspections.ipynb`

The goal of `A_EvaluationMetrics.ipynb` is to compute the most relevant metrics from the `02_train` predictions and compares the newer, LSTM predictions to the process-based ice flag performance. Relevant metrics include the accuracy of ice prediction on all days, the predicted max ice on date, the predicted max ice off date, and the predicted ice duration.

These errors/metrics are saved for further inspection.

`B_PlottingInspections.ipynb` uses the saved errors/metrics to perform some further, plot-based evaluation. Right now, this is exclusive to visualizing some time series and the spatial distribution of errors.

# Next steps

Prior to adding a decked out LSTM, I will be experimenting with the variable set to hopefully omit geographically-limiting variables (e.g., clarity, max depth temperature estimates) and add additional promising variables (e.g., lake surface area). 

The advanced LSTM efforts will be next. This may include some low-hanging learning rate improvements, loss function customization to emphasize the transition periods of the time series, and potentially PB pretraining. The first two have appeared promising in prototyping but have not been polished, the third is untouched but easily accessible. 

# Requested review

May include but not limited to:

* General bug identification
* Ideas for advanced LSTM model
* Ideas for further evaluation (caveat: I think the specific results of this first pass LSTM should not be dwelled on too much yet)

### CODE REVIEW

**Review by AndyMcAliley — CHANGES_REQUESTED** [review]
Nice work! I'm surprised how well the model predicts, and how many epochs you can train for and see improvement, especially with only one hidden dimension. It doesn't look like there's much overfitting. It looks like ice-off is easier to predict than ice-on - interesting!

## Code comments

The only part of the code I have comments about is in "Derive predicted max ice on and max ice off dates" in `A_EvaluationMetrics.ipynb`. It looks like you're using the final fall freeze and the final spring thaw. Is that correct, and is that what you want?

`ice_on_index` is based on the diff array, so it's one less than the index of the day when ice appears. Is this causing an off-by-one error in `ice_on`, `ice_off`, `pred_ice_on`, `pred_ice_off`, `flag_ice_on`,  and `flag_ice_off`?

Also, I'm not sure, but lakes that do not freeze at all might result in errors in this part of the code. Something to keep in mind if you expand beyond Minnesota, or for projections under climate warming.

## Thoughts

- Is there a need for model spin-up time? If you do try mixing up the start date, then there probably will be such a need.
- Might be useful to make a scatter plot of predicted ice-on day vs observed ice-on day to see how well outlier years are modeled.

**Comment by jdiaz4302** [comment]
> The only part of the code I have comments about is in "Derive predicted max ice on and max ice off dates" in A_EvaluationMetrics.ipynb. It looks like you're using the final fall freeze and the final spring thaw. Is that correct, and is that what you want?

Yeah, these were the dates that Gretchen and Jordan suggested using as most relevant. 

> ice_on_index is based on the diff array, so it's one less than the index of the day when ice appears. Is this causing an off-by-one error in ice_on, ice_off, pred_ice_on, pred_ice_off, flag_ice_on, and flag_ice_off?

Good catch, I'll add that first index back into the diff array. It doesn't cause an error because I'm using those diff arrays to find a singular transition-date index, but you're still right that it should be fixed to give more accurate evaluation.

> Also, I'm not sure, but lakes that do not freeze at all might result in errors in this part of the code. Something to keep in mind if you expand beyond Minnesota, or for projections under climate warming.

The data set is constructed for ice observations, so every sequence is required to have an icy period to have a sequence made. I didn't include any ice-free sequences because it's not certain if a year without ice observations is truly lacking ice or if it's just lacking observation; given the data set is Minnesota, I believe lacking observation is the right answer.

Definitely a good point for later evaluation, that we'll should be able to handle a sequence with all-zero diff array and, also, probably want to evaluate on some ice-free data.

> Is there a need for model spin-up time? If you do try mixing up the start date, then there probably will be such a need.

I'll add this note to the issue on different starting dates for sequences.

> Might be useful to make a scatter plot of predicted ice-on day vs observed ice-on day to see how well outlier years are modeled.

I think this same information is captured in the histograms generated by `plot_and_print_date_errors` in `A_EvaluationMetrics.ipynb`

**Comment by AndyMcAliley** [comment]
Code changes sound good!

> > Might be useful to make a scatter plot of predicted ice-on day vs observed ice-on day to see how well outlier years are modeled.
> 
> I think this same information is captured in the histograms generated by plot_and_print_date_errors in A_EvaluationMetrics.ipynb

Here's an example of what I had in mind (plotting flagged ice-on days, but could sub in predicted ice-on days from the LSTM for flagged days):

test_scatter

The scatter plot shows that flagged ice-on days are early, like the histogram shows. It also shows that flagged ice-on days tend to agree less well with later observed ice-on days, which is information that the histogram doesn't convey.

File: 03_eval/src/A_EvaluationMetrics.ipynb [hunk 1]
```diff
@@ -322,9 +327,9 @@
     "\n",
     "for i in range(valid_y.shape[0]):\n",
     "    # Result in +1 when ice emerges and -1 when ice disappears. 0 otherwise.\n",
-    "    valid_diff_y_i = np.diff(valid_y[i, :])\n",
-    "    valid_diff_y_hat_i = np.diff(valid_predictions_class[i, :, 0])\n",
-    "    valid_diff_flag_i = np.diff(valid_x[i, :, ice_var_idx])\n",
+    "    valid_diff_y_i = np.concatenate([np.asarray([0]), np.diff(valid_y[i, :])])\n",
+    "    valid_diff_y_hat_i = np.concatenate([np.asarray([0]), np.diff(valid_predictions_class[i, :, 0])])\n",
+    "    valid_diff_flag_i = np.concatenate([np.asarray([0]), np.diff(valid_x[i, :, ice_var_idx])])\n",
```
Comment (jdiaz4302) [inline 03_eval/src/A_EvaluationMetrics.ipynb:329]: This takes care of the `np.diff` issue. To reintroduce that first index that the differencing removes, I just add in a 0-value at the beginning. Here, a zero value indicates no change which is physically realistic for the state of ice on July 1.

Notably, this did not change any of the evaluated metrics because previous **ALL** dates were off by an index of 1 - observations, PB ice flags, and LSTM predictions.

File: 03_eval/src/A_EvaluationMetrics.ipynb [hunk 2]
```diff
@@ -298,21 +301,23 @@
...
    "outputs": [],
    "source": [
+    "# These objects are easier to do basic math with\n",
+    "# because they are date indices\n",
+    "ice_on_ids = []\n",
+    "ice_off_ids = []\n",
+    "pred_ice_on_ids = []\n",
+    "pred_ice_off_ids = []\n",
+    "flag_ice_on_ids = []\n",
+    "flag_ice_off_ids = []\n",
```
Comment (jdiaz4302) [inline 03_eval/src/A_EvaluationMetrics.ipynb:315]: New objects to more easily generate the requested plots using indices (i.e., days after July 1) rather than literal dates

File: 03_eval/src/A_EvaluationMetrics.ipynb [hunk 3]
```diff
@@ -576,10 +596,172 @@
...
+   "id": "d268e949",
+   "metadata": {},
+   "source": [
+    "<br><br><br>\n",
+    "\n",
+    "# Additional recommended plots\n",
+    "\n",
+    "I'm doing these in the evaluation metrics notebook instead of the plotting notebook because they are related to higher level prediction targets considered here, rather than evaluating residuals, which is the primary focus of the other notebook.\n",
+    "\n",
+    "### Scatter plots of observations versus predictions"
```
Comment (jdiaz4302) [inline 03_eval/src/A_EvaluationMetrics.ipynb:607]: Requested plots start here (along with rationale for why in this notebook)

**Comment by jdiaz4302** [comment]
Good plot suggestions

In all cases the LSTM predictions hug the 1:1 better.

In ice on and ice off, you can see that it corrects _most_ of the early bias and the worsening trend with later dates (but not every outlier case).

These also demonstrate that the PB models don't have as much of a bias when it comes to duration (they seem to approximately get the duration right while just shifting everything early). Ultimately, LSTM still hugs the 1:1 better.

**Comment by AndyMcAliley** [comment]
Good idea to make that duration scatter plot! 

I agree with your observations. I'd also comment that both models tend to overestimate early ice on/off days, and underestimate late ice on/off days, but the LSTM over/underestimates less that the process-based model. Very interesting - I think of process-based models as being more reliable for estimating outliers that aren't well-represented in the data, but here the LSTM is outperforming the PB model on outliers, too!

**Review by AndyMcAliley — APPROVED** [review]
